File: apps/web/lib/notifications-link.ts
/**
 * Cible de navigation d'une notification — déduite du template et de ses variables
 * (contenu_json), côté client comme côté serveur. Une notification cliquée ouvre l'objet
 * concerné (AG, lot, incident…), jamais une page générique quand l'objet est connu.
 */
export function lienNotification(
  templateCode: string,
  contenu: Record<string, unknown> | null | undefined,
  locale: string
): string {
  const c = contenu ?? {};
  const id = (k: string) => (typeof c[k] === "string" ? (c[k] as string) : null);
  const p = (path: string) => `/${locale}${path}`;

  if (templateCode === "PV_DISPONIBLE" && id("ag_id")) return p(`/ag/${id("ag_id")}/pv`);
  if (templateCode.startsWith("AG_") && id("ag_id")) return p(`/ag/${id("ag_id")}`);
  if (templateCode.startsWith("INCIDENT_") && id("incident_id")) {
    return p(`/incidents/${id("incident_id")}`);
  }
  if (
    templateCode === "APPEL_DE_FONDS_EMIS" ||
    templateCode === "PAIEMENT_RECU" ||
    templateCode.startsWith("IMPAYE_")
  ) {
    if (id("lot_id")) return p(`/lots/${id("lot_id")}?onglet=finances`);
    if (id("appel_de_fonds_id")) return p(`/finances/appels-de-fonds/${id("appel_de_fonds_id")}`);
    return p("/finances/appels-de-fonds");
  }
  if (templateCode.startsWith("CONTESTATION_")) return p("/finances/contestations");
  if (templateCode === "LOT_RATTACHE" && id("lot_id")) return p(`/lots/${id("lot_id")}`);
  if (templateCode.startsWith("DOCUMENT_")) return p("/documents");
  if (templateCode === "INVITATION_ACCEPTEE") {
    return id("utilisateur_id") ? p(`/membres/${id("utilisateur_id")}`) : p("/invitations");
  }
  if (templateCode.startsWith("VISITE_")) return p("/visites");
  if (templateCode.startsWith("RESERVATION_")) return p("/reservations");
  if (templateCode.startsWith("LITIGE_")) return p("/litiges");
  return p("/notifications");
}

/** Événement DOM émis quand une notification passe « lue » côté client (cloche, listes). */
export const NOTIF_LUE_EVENT = "su:notif-lue";

export function signalerLue(id: string) {
  if (typeof window === "undefined") return;
  window.dispatchEvent(new CustomEvent(NOTIF_LUE_EVENT, { detail: { id } }));
}

/** Marquage « lu » en arrière-plan (proxy web → API). Ne bloque jamais l'interface. */
export function marquerLueEnFond(id: string): void {
  if (typeof window === "undefined") return;
  signalerLue(id);
  void fetch("/api/notifications-read", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ id }),
    keepalive: true,
  }).catch(() => {
    /* réseau : la prochaine synchronisation rattrapera l'état */
  });
}

File: apps/mobile/lib/core/util/notifications_link.dart
/// Cible de navigation d'une notification (push ou in-app) — déduite du template et de ses
/// variables. Même table que apps/web/lib/notifications-link.ts (deep-links brief §8.2).
String lienNotification(String templateCode, Map<String, dynamic>? contenu) {
  final c = contenu ?? const {};
  String? id(String k) => c[k] is String ? c[k] as String : null;

  if (templateCode == 'PV_DISPONIBLE' && id('ag_id') != null) return '/ag/${id('ag_id')}/pv';
  if (templateCode.startsWith('AG_') && id('ag_id') != null) return '/ag/${id('ag_id')}';
  if (templateCode.startsWith('INCIDENT_') && id('incident_id') != null) return '/incidents/${id('incident_id')}';
  if (templateCode == 'APPEL_DE_FONDS_EMIS' || templateCode == 'PAIEMENT_RECU' || templateCode.startsWith('IMPAYE_')) {
    if (id('lot_id') != null) return '/lots/${id('lot_id')}?onglet=finances';
    if (id('appel_de_fonds_id') != null) return '/finances/appels-de-fonds/${id('appel_de_fonds_id')}';
    return '/finances/appels-de-fonds';
  }
  if (templateCode.startsWith('CONTESTATION_')) return '/finances/contestations';
  if (templateCode == 'LOT_RATTACHE' && id('lot_id') != null) return '/lots/${id('lot_id')}';
  if (templateCode.startsWith('DOCUMENT_')) return '/documents';
  if (templateCode == 'INVITATION_ACCEPTEE') return id('utilisateur_id') != null ? '/membres/${id('utilisateur_id')}' : '/invitations';
  if (templateCode.startsWith('VISITE_')) return id('visite_id') != null ? '/visites/${id('visite_id')}' : '/visites';
  if (templateCode.startsWith('RESERVATION_')) return '/reservations';
  if (templateCode.startsWith('LITIGE_')) return '/litiges';
  // M17 — justificatifs : déclaration à valider, validation, rejet, espèces, relance.
  if (templateCode.startsWith('JUSTIFICATIF_') || templateCode == 'PAIEMENT_VALIDE' || templateCode == 'PAIEMENT_ESPECES_SAISI') {
    return id('justificatif_id') != null ? '/justificatifs/${id('justificatif_id')}' : '/justificatifs';
  }
  // M19 — contrats : échéances, expiration, reconduction, assurance absente.
  if (templateCode.startsWith('CONTRAT_')) return id('contrat_id') != null ? '/contrats/${id('contrat_id')}' : '/contrats';
  if (templateCode == 'ASSURANCE_IMMEUBLE_ABSENTE') return '/contrats';
  // M18 — rapport de gestion soumis à l'AG : les copropriétaires le lisent depuis la transparence.
  if (templateCode.startsWith('RAPPORT_GESTION_')) return '/rapports/transparence';
  // M16 — dépenses : approbation, décision, échéance de facture.
  if (templateCode.startsWith('DEPENSE_') || templateCode == 'FACTURE_ECHEANCE_PROCHE') {
    return id('depense_id') != null ? '/depenses/${id('depense_id')}' : '/depenses';
  }
  if (templateCode.startsWith('LCD_')) {
    if (id('sejour_id') != null) return '/location-courte-duree/sejours/${id('sejour_id')}';
    if (id('declaration_id') != null) return '/location-courte-duree/declarations/${id('declaration_id')}';
    return '/location-courte-duree';
  }
  return '/notifications';
}
